src/recipe/parser/source.rs

```diff
@@ -0,0 +1,409 @@
+                    ))
+                }
+            }
+        }
+
+        let url = url.ok_or_else(|| {
+            _partialerror!(
+                *self.span(),
+                ErrorKind::MissingField("git_url".into()),
+                help = "git `source` must have a `git_url` field"
+            )
+        })?;
+
+        let rev = rev.unwrap_or_else(|| "HEAD".to_owned());
+
+        Ok(GitSource {
+            url,
+            rev,
+            depth,
+            patches,
+            folder,
+        })
+    }
+}
+
+/// git url
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum GitUrl {
+    Url(Url),
+    Path(PathBuf),
+}
+
+impl fmt::Display for GitUrl {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            GitUrl::Url(url) => write!(f, "{url}"),
+            GitUrl::Path(path) => write!(f, "{path:?}"),
+        }
+    }
+}
+
+/// A url source (usually a tar.gz or tar.bz2 archive). A compressed file
+/// will be extracted to the `work` (or `work/<folder>` directory).
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct UrlSource {
+    /// Url to the source code (usually a tar.gz or tar.bz2 etc. file)
+    url: Url,
+    /// Optionally a checksum to verify the downloaded file
+    checksums: Vec<Checksum>,
+    /// Optionally a file name to rename the downloaded file (does not apply to archives)
+    file_name: Option<String>,
+    /// Patches to apply to the source code
+    patches: Vec<PathBuf>,
+    /// Optionally a folder name under the `work` directory to place the source code
+    folder: Option<PathBuf>,
+}
+
+impl UrlSource {
+    /// Get the url.
+    pub const fn url(&self) -> &Url {
+        &self.url
+    }
+
+    /// Get the checksum of the URL source.
+    pub fn checksums(&self) -> &[Checksum] {
+        self.checksums.as_slice()
+    }
+
+    /// Get the patches of the URL source.
+    pub fn patches(&self) -> &[PathBuf] {
+        self.patches.as_slice()
+    }
+
+    /// Get the folder of the URL source.
+    pub const fn folder(&self) -> Option<&PathBuf> {
+        self.folder.as_ref()
+    }
+
+    /// Get the file name of the URL source.
+    pub const fn file_name(&self) -> Option<&String> {
+        self.file_name.as_ref()
+    }
+}
+
+impl TryConvertNode<UrlSource> for RenderedMappingNode {
+    fn try_convert(&self, _name: &str) -> Result<UrlSource, PartialParsingError> {
+        let mut url = None;
+        let mut checksums = Vec::new();
+        let mut patches = Vec::new();
+        let mut folder = None;
+        let mut file_name = None;
+
+        for (key, value) in self.iter() {
+            match key.as_str() {
+                "url" => url = Some(value.try_convert("url")?),
+                "sha256" => {
+                    let sha256_str: String = value.try_convert("sha256")?;
+                    if sha256_str.len() != 64 {
```

Comment:
There are actually some more constraints. Maybe we could use `rattler_digest::Md5Hash` and `rattler_digest::Sha256Hash` that would parse it to the right types straight away?

Reply:
That is a good idea!